k8s/upgrade/src/bin/upgrade-job/helm/values.rs

```diff
@@ -203,3 +249,70 @@ where
 
     Ok(upgrade_values_file)
 }
+
+/// Converts config.lokiAddress and config.snippets.extraClientConfigs from the promtail helm chart
+/// v3.11.0 to config.clients[] which is compatible with promtail helm chart v6.13.1.
+fn loki_address_to_clients(
+    target_values: &CoreValues,
+    upgrade_values_filepath: &Path,
+    yq: &YqV4,
+) -> Result<()> {
+    let promtail_config_clients_yaml_key =
+        YamlKey::try_from(".loki-stack.promtail.config.clients")?;
+    // Delete existing array, if any. The merge_files() should have added it with the default value
+    // set.
+    yq.delete_object(
+        promtail_config_clients_yaml_key.clone(),
+        upgrade_values_filepath,
+    )?;
+    let loki_address = target_values.loki_stack_promtail_loki_address();
+    let promtail_config_client = PromtailConfigClient::with_url(loki_address);
+    let promtail_config_client = serde_json::to_string(&promtail_config_client).context(
+        SerializePromtailConfigClientToJson {
+            object: promtail_config_client,
+        },
+    )?;
+    yq.append_to_array(
+        promtail_config_clients_yaml_key,
+        promtail_config_client,
+        upgrade_values_filepath,
+    )?;
+    // Merge the extraClientConfigs from the promtail v3.11.0 chart to the v6.13.1 chart's
+    // config.clients block. Ref: https://github.com/grafana/helm-charts/issues/1214
+    // Ref: https://github.com/grafana/helm-charts/pull/1425
+    if !target_values.promtail_extra_client_configs().is_empty() {
+        // Converting the YAML to a JSON because the yq command goes like this...
+        // yq '.config.clients[0] += {"tenant_id": "1", "basic_auth": {"username": "loki",
+        // "password": "secret"}}' file
```

Comment:
Don't understand this comment..
Did you mean to say yq works with a json syntax? If so it'd be clearer to just say that..